eras/mary/impl/src/Cardano/Ledger/Mary/Value.hs

```diff
@@ -262,58 +265,85 @@ roundupBytesToWords b = quot (b + wordLength - 1) wordLength
 -- Maybe better to make this distinction in the TxOut de/serialization
 
 decodeValue ::
+  forall c s.
   Crypto c =>
   Decoder s (MaryValue c)
-decodeValue = do
-  tt <- peekTokenType
-  case tt of
-    TypeUInt -> inject . Coin <$> decodeInteger
-    TypeUInt64 -> inject . Coin <$> decodeInteger
-    TypeNInt -> inject . Coin <$> decodeInteger
-    TypeNInt64 -> inject . Coin <$> decodeInteger
-    TypeListLen -> decodeValuePair decodeInteger
-    TypeListLen64 -> decodeValuePair decodeInteger
-    TypeListLenIndef -> decodeValuePair decodeInteger
-    _ -> fail $ "Value: expected array or int, got " ++ show tt
+decodeValue =
+  ifDecoderVersionAtLeast
+    (natVersion @9)
+    (decode' conwayCoin)
+    (decode' preConwayCoin)
+  where
+    conwayCoin :: Decoder s (MaryValue c)
+    conwayCoin = inject . word64ToCoin <$> decodeWord64
+    preConwayCoin :: Decoder s (MaryValue c)
+    preConwayCoin = inject . Coin <$> decodeInteger
```

Comment:
`inject . word64ToCoin <$> decodeWord64` here too. But for Conway guard against zero too.